File: eval-score/node/src/evalset-loader.ts
import * as fs from 'fs';
import * as path from 'path';
import { EvalRow, Assertion } from './types';

/**
 * EvalGen EvalSet JSON format (mirrors eval-gen/src/types.ts EvalSet)
 */
interface EvalSetJson {
  version: string;
  generated_at: string;
  description: string;
  source_file: string;
  item_count: number;
  items: Array<{
    id?: string;
    prompt: string;
    expected_answer: string;
    source_location: string;
    assertions?: Assertion[];
    category?: string;
    difficulty?: string;
    grounding_confidence?: string;
  }>;
  warnings?: string[];
  metadata?: {
    model?: string;
    evalgen_version?: string;
  };
}

/**
 * Load an EvalGen EvalSet JSON directly as EvalRows with assertions pre-attached.
 * This is the preferred integration path — no need for separate --sidecar.
 */
export function loadEvalSet(evalsetPath: string): {
  rows: EvalRow[];
  warnings: string[];
  metadata: Record<string, string>;
} {
  const absPath = path.resolve(evalsetPath);
  if (!fs.existsSync(absPath)) {
    throw new Error(`EvalSet file not found: ${absPath}`);
  }

  const content = fs.readFileSync(absPath, 'utf-8');
  let evalSet: EvalSetJson;
  try {
    evalSet = JSON.parse(content);
  } catch {
    throw new Error(`Invalid JSON in EvalSet file: ${absPath}`);
  }

  if (!evalSet.items || !Array.isArray(evalSet.items)) {
    throw new Error('Invalid EvalSet format: missing "items" array');
  }

  if (evalSet.version && !evalSet.version.startsWith('1.')) {
    process.stderr.write(`  ⚠️  EvalSet version ${evalSet.version} may not be compatible\n`);
  }

  const rows: EvalRow[] = evalSet.items.map(item => ({
    prompt: item.prompt,
    expectedAnswer: item.expected_answer,
    sourceLocation: item.source_location,
    actualAnswer: '',
    assertions: item.assertions,
  }));

  // Collect metadata for reporting
  const metadata: Record<string, string> = {};
  if (evalSet.description) metadata.description = evalSet.description;
  if (evalSet.source_file) metadata.source_file = evalSet.source_file;
  if (evalSet.generated_at) metadata.generated_at = evalSet.generated_at;
  if (evalSet.metadata?.model) metadata.model = evalSet.metadata.model;
  if (evalSet.metadata?.evalgen_version) metadata.evalgen_version = evalSet.metadata.evalgen_version;

  // Category metadata for per-category reporting
  for (let i = 0; i < rows.length && i < evalSet.items.length; i++) {
    const item = evalSet.items[i];
    // Store category/difficulty as non-enumerable properties for reporting
    (rows[i] as any)._category = item.category;
    (rows[i] as any)._difficulty = item.difficulty;
    (rows[i] as any)._confidence = item.grounding_confidence;
    (rows[i] as any)._id = item.id;
  }

  return {
    rows,
    warnings: evalSet.warnings ?? [],
    metadata,
  };
}

File: eval-ui-winui3/src/EvalToolkit.EvalScore/EvalSet/EvalSetLoader.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using EvalToolkit.EvalScore.Models;

namespace EvalToolkit.EvalScore.EvalSet;

/// <summary>
/// Loader for the EvalGen <c>*.evalgen.json</c> EvalSet format. Mirrors
/// TS <c>loadEvalSet</c> in <c>eval-score/node/src/evalset-loader.ts</c>.
///
/// <para>Behavior preserved:
/// <list type="bullet">
///   <item>Path is resolved to absolute before existence check.</item>
///   <item>Missing file throws <see cref="FileNotFoundException"/>.</item>
///   <item>Invalid JSON throws <see cref="InvalidOperationException"/>
///     with the absolute path in the message.</item>
///   <item>Missing or non-array <c>items</c> property throws
///     <see cref="InvalidOperationException"/>.</item>
///   <item>Version not starting with <c>"1."</c> emits a warning via
///     <see cref="OnVersionWarning"/> (the TS code writes to stderr).</item>
///   <item>Each item maps to a fresh <see cref="EvalRow"/> with empty
///     <see cref="EvalRow.ActualAnswer"/>; assertions, category,
///     difficulty, grounding_confidence, and id are copied through.</item>
///   <item>Metadata (description, source_file, generated_at, model,
///     evalgen_version) is collected into a string→string dict for
///     reporters to use.</item>
/// </list></para>
/// </summary>
public static class EvalSetLoader
{
    private static readonly JsonSerializerOptions s_jsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.SnakeCaseLower) },
    };

    /// <summary>Optional sink for version-mismatch warnings (defaults to <see cref="Console.Error"/>).</summary>
    public static Action<string>? OnVersionWarning { get; set; }

    public static LoadResult Load(string evalsetPath)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(evalsetPath);
        string absPath = Path.GetFullPath(evalsetPath);
        if (!File.Exists(absPath))
        {
            throw new FileNotFoundException($"EvalSet file not found: {absPath}", absPath);
        }

        string content = File.ReadAllText(absPath);
        EvalSetJson evalSet;
        try
        {
            evalSet = JsonSerializer.Deserialize<EvalSetJson>(content, s_jsonOptions)
                ?? throw new InvalidOperationException($"Invalid JSON in EvalSet file: {absPath}");
        }
        catch (JsonException)
        {
            throw new InvalidOperationException($"Invalid JSON in EvalSet file: {absPath}");
        }

        if (evalSet.Items is null)
        {
            throw new InvalidOperationException("Invalid EvalSet format: missing \"items\" array");
        }

        if (!string.IsNullOrEmpty(evalSet.Version) && !evalSet.Version.StartsWith("1.", StringComparison.Ordinal))
        {
            string warning = $"  ⚠️  EvalSet version {evalSet.Version} may not be compatible";
            if (OnVersionWarning is not null)
            {
                OnVersionWarning(warning);
            }
            else
            {
                Console.Error.WriteLine(warning);
            }
        }

        var rows = new List<EvalRow>(evalSet.Items.Count);
        foreach (EvalSetItemJson item in evalSet.Items)
        {
            rows.Add(new EvalRow
            {
                Prompt = item.Prompt,
                ExpectedAnswer = item.ExpectedAnswer,
                SourceLocation = item.SourceLocation,
                ActualAnswer = string.Empty,
                Assertions = item.Assertions,
                // TS evalset-loader stores item.id as (row as any)._id (a
                // separate non-`id` field that no consumer ever reads after
                // assignment). Mirroring that behavior here keeps checkpoint
                // output free of an unexpected extensions.evalscore.item_id
                // for EvalSet-loaded rows. Category/Difficulty/
                // GroundingConfidence DO have downstream readers (reporter
                // per-category grouping) and are preserved.
                Category = item.Category,
                Difficulty = item.Difficulty,
                GroundingConfidence = item.GroundingConfidence,
            });
        }

        var metadata = new Dictionary<string, string>();
        if (!string.IsNullOrEmpty(evalSet.Description)) metadata["description"] = evalSet.Description;
        if (!string.IsNullOrEmpty(evalSet.SourceFile)) metadata["source_file"] = evalSet.SourceFile;
        if (!string.IsNullOrEmpty(evalSet.GeneratedAt)) metadata["generated_at"] = evalSet.GeneratedAt;
        if (!string.IsNullOrEmpty(evalSet.Metadata?.Model)) metadata["model"] = evalSet.Metadata.Model;
        if (!string.IsNullOrEmpty(evalSet.Metadata?.EvalGenVersion)) metadata["evalgen_version"] = evalSet.Metadata.EvalGenVersion;

        return new LoadResult(rows, evalSet.Warnings ?? new List<string>(), metadata);
    }

    public sealed record LoadResult(IList<EvalRow> Rows, IList<string> Warnings, IDictionary<string, string> Metadata);

    private sealed class EvalSetJson
    {
        public string? Version { get; set; }
        public string? GeneratedAt { get; set; }
        public string? Description { get; set; }
        public string? SourceFile { get; set; }
        public int ItemCount { get; set; }
        public List<EvalSetItemJson>? Items { get; set; }
        public List<string>? Warnings { get; set; }
        public EvalSetMetadataJson? Metadata { get; set; }
    }

    private sealed class EvalSetItemJson
    {
        public string? Id { get; set; }
        public string Prompt { get; set; } = string.Empty;
        public string ExpectedAnswer { get; set; } = string.Empty;
        public string SourceLocation { get; set; } = string.Empty;
        public List<Assertion>? Assertions { get; set; }
        public string? Category { get; set; }
        public string? Difficulty { get; set; }
        public string? GroundingConfidence { get; set; }
    }

    private sealed class EvalSetMetadataJson
    {
        public string? Model { get; set; }
        [JsonPropertyName("evalgen_version")]
        public string? EvalGenVersion { get; set; }
    }
}
